src/meta/api/src/schema_api_impl.rs

```diff
@@ -1783,6 +1784,147 @@ impl<KV: kvapi::KVApi<Error = KVAppError>> SchemaApi for KV {
+            };
+
+            let tbname = dbid_tbname.table_name.clone();
+            let (tb_id_seq, _) = get_u64_value(self, &dbid_tbname).await?;
+            if tb_id_seq == 0 {
+                return Err(KVAppError::AppError(AppError::UnknownTableId(
+                    UnknownTableId::new(table_id, "drop_table_by_id failed to get valid tb_id_seq"),
+                )));
+            }
+
+            let db_id_to_name = DatabaseIdToName {
+                db_id: dbid_tbname.db_id,
+            };
+            let (_, database_name_opt): (_, Option<DatabaseNameIdent>) =
+                get_struct_value(self, &db_id_to_name).await?;
+            let tenant_dbname = match database_name_opt {
+                Some(db_name_ident) => db_name_ident,
+                None => {
+                    return Err(KVAppError::AppError(AppError::UnknownDatabaseId(
+                        UnknownDatabaseId::new(dbid_tbname.db_id, "drop_table_by_id"),
+                    )));
+                }
+            };
+            let tenant_dbname_tbname = TableNameIdent {
+                tenant: tenant_dbname.tenant.clone(),
+                db_name: tenant_dbname.db_name.clone(),
+                table_name: tbname,
+            };
+
+            // get current table count from _fd_table_count/tenant
+            let tb_count_key = CountTablesKey {
+                tenant: tenant_dbname.tenant.clone(),
+            };
+            (tb_count_seq, tb_count) = {
+                let (seq, count) = get_u64_value(self, &tb_count_key).await?;
+                if seq > 0 {
+                    (seq, count)
+                } else if !tbcount_found {
+                    // only count_tables for the first time.
+                    tbcount_found = true;
+                    (0, count_tables(self, &tb_count_key).await?)
+                } else {
+                    (0, tb_count)
+                }
+            };
+
+            let (_, db_id, db_meta_seq, db_meta) =
+                get_db_or_err(self, &tenant_dbname, "drop_table_by_id").await?;
+
+            debug!(
+                ident = display(&tbid),
+                name = display(&tenant_dbname_tbname),
+                "drop table by id"
+            );
+
+            {
+                let mut tb_meta = tb_meta.unwrap();
+                // drop a table with drop_on time
+                if tb_meta.drop_on.is_some() {
+                    return Err(KVAppError::AppError(AppError::DropTableWithDropTime(
+                        DropTableWithDropTime::new(&dbid_tbname.table_name),
+                    )));
+                }
+
+                tb_meta.drop_on = Some(Utc::now());
+
+                let txn_req = TxnRequest {
+                    condition: vec![
+                        // db has not to change, i.e., no new table is created.
+                        // Renaming db is OK and does not affect the seq of db_meta.
+                        txn_cond_seq(&DatabaseId { db_id }, Eq, db_meta_seq),
```

Comment:
IMHO, dropping a table does not involve any changes to its belonging database.
Is it necessary to assert the database-meta not changed?

Reply:
Thanks for taking a look! I think it's is. It's from https://github.com/datafuselabs/databend/pull/5186/files. create/drop/rename a table will cause a db change. Otherwise, if the table is renamed after we receive the request and before we commit the drop transaction, some metadata we get may be staled. Then the question is, do we still want to proceed on this request when there's a table update on db?

Reply:
Oh right. there is a table-name list for every db.